imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,334 @@
+    """
+    if output_map_properties.sky_tiling_type is ena_maps.SkyTilingType.HEALPIX:
+        map_nside, map_nested = (
+            output_map_properties.nside,
+            output_map_properties.nested,
+        )
+    else:
+        map_nside, map_nested = (DEFAULT_L2_HEALPIX_NSIDE, DEFAULT_L2_HEALPIX_NESTED)
+
+    # Initialize the HealpixSkyMap object
+    skymap = ena_maps.HealpixSkyMap(
+        nside=map_nside,
+        nested=map_nested,
+        spice_frame=output_map_properties.spice_reference_frame,
+    )
+
+    # Add additional data variables to the map
+    output_map_properties.values_to_push_project.extend(
+        [
+            "observation_time",
+            "pointing_set_exposure_times_solid_angle",
+            "num_pointing_set_pixel_members",
+        ]
+    )
+
+    for ultra_l1c_pset in ultra_l1c_psets:
+        pointing_set = read_into_pointing_set(ultra_l1c_pset)
+        logger.info(f"PSET epoch: {pointing_set.epoch}")
+
+        pointing_set.data["num_pointing_set_pixel_members"] = xr.DataArray(
+            np.ones(pointing_set.num_points, dtype=int),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        pointing_set.data["observation_time"] = xr.DataArray(
+            np.full(pointing_set.num_points, pointing_set.epoch),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        # Add solid_angle * exposure of pointing set as data_var
+        # so this quantity is projected to map pixels for use in weighted averaging
+        pointing_set.data["pointing_set_exposure_times_solid_angle"] = (
+            pointing_set.data["exposure_factor"] * pointing_set.solid_angle
+        )
+
+        # Initial processing for weighted quantities at PSET level
+        # Weight the values by exposure and solid angle
+        # (in that order to avoid double weighting by solid angle)
+        for (
+            quantity_to_weight
+        ) in VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE:
+            pointing_set.data[quantity_to_weight] = (
+                pointing_set.data[quantity_to_weight]
+                * pointing_set.data["pointing_set_exposure_times_solid_angle"]
+            )
+
+        skymap.project_pset_values_to_map(
+            pointing_set=pointing_set,
+            value_keys=list(
+                set(
+                    output_map_properties.values_to_push_project
+                    + REQUIRED_L1C_VARIABLES
+                )
+            ),
+            index_match_method=ena_maps.IndexMatchMethod.PUSH,
+        )
+
+    # Subsequent processing for weighted quantities at SkyMap level
+    for (
+        quantity_to_weight
+    ) in VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE:
+        skymap.data_1d[quantity_to_weight] = (
+            skymap.data_1d[quantity_to_weight]
+            / skymap.data_1d["pointing_set_exposure_times_solid_angle"]
+        )
+
+    # TODO: Ask Ultra team about this - I think this is a decent
```

Comment:
This will potentially be incorrect if one pointing lasted 3 days and another pointing only lasted a half-day or things like that where the view-time per pixel would be wildly different between the pointing sets I think. Do we need to take into account time here as well?

Reply:
I think you're right. I'll note this as well and bring it up ASAP with Ultra, though I'm not sure progress will happen until Nick is back from vacation. I think that adding the "*mean-over-a-PSET-but-sum-over-different-PSETs*" functionality I've been thinking about to the ena_maps code will fix this as well but I need some more info from the Ultra team.

Reply:
I'll make a ticket for now.